iceoryx2/src/sample_mut.rs

```diff
@@ -12,91 +12,152 @@
-//! let sample = sample.write_payload(1234);
+//! let sample = write_sample_data(sample);
 //!
-//! println!("timestamp: {:?}, publisher port id: {:?}",
-//!     sample.header().time_stamp(), sample.header().publisher_id());
-//! publisher.send(sample)?;
+//! sample.send()?;
 //!
 //! # Ok(())
 //! # }
 //! ```
+//! See also, [`crate::sample_mut_impl::SampleMutImpl`].
 
-use crate::{port::publisher::Publisher, raw_sample::RawSampleMut, service};
-use iceoryx2_cal::shared_memory::*;
-use std::{fmt::Debug, mem::MaybeUninit, sync::atomic::Ordering};
+use iceoryx2_cal::zero_copy_connection::ZeroCopyCreationError;
 
-/// Acquired by a [`Publisher`] via [`Publisher::loan()`] or [`Publisher::loan_uninit()`]. It stores the payload that will be sent
-/// to all connected [`crate::port::subscriber::Subscriber`]s. If the [`SampleMut`] is not sent
-/// it will release the loaned memory when going out of scope.
-///
-/// # Notes
-///
-/// Does not implement [`Send`] since it releases unsent samples in the [`Publisher`] and the
-/// [`Publisher`] is not thread-safe!
-///
-/// The generic parameter `M` is either a `MessageType` or a [`core::mem::MaybeUninit<MessageType>`], depending
-/// which API is used to obtain the sample.
-#[derive(Debug)]
-pub struct SampleMut<
-    'a,
-    'publisher,
-    'config,
-    Service: service::Details<'config>,
-    Header: Debug,
-    M: Debug,
-> {
-    publisher: &'publisher Publisher<'a, 'config, Service, M>,
-    ptr: RawSampleMut<Header, M>,
-    offset_to_chunk: PointerOffset,
-}
+use crate::service::header::publish_subscribe::Header;
 
-impl<'config, Service: service::Details<'config>, Header: Debug, M: Debug> Drop
-    for SampleMut<'_, '_, 'config, Service, Header, M>
-{
-    fn drop(&mut self) {
-        self.publisher.release_sample(self.offset_to_chunk);
-        self.publisher.loan_counter.fetch_sub(1, Ordering::Relaxed);
+pub(crate) mod internal {
+    use iceoryx2_cal::zero_copy_connection::PointerOffset;
+
+    pub trait SampleMgmt {
+        fn offset_to_chunk(&self) -> PointerOffset;
     }
 }
 
-impl<
-        'a,
-        'publisher,
-        'config,
-        Service: service::Details<'config>,
-        Header: Debug,
-        MessageType: Debug,
-    > SampleMut<'a, 'publisher, 'config, Service, Header, MaybeUninit<MessageType>>
-{
-    pub(crate) fn new(
-        publisher: &'publisher Publisher<'a, 'config, Service, MessageType>,
-        ptr: RawSampleMut<Header, MaybeUninit<MessageType>>,
-        offset_to_chunk: PointerOffset,
-    ) -> Self {
-        publisher.loan_counter.fetch_add(1, Ordering::Relaxed);
+/// Acquired by a [`crate::port::publisher::Publisher`] via
+/// [`crate::port::publisher::PublisherLoan::loan()`]. It stores the payload that will be sent
+/// to all connected [`crate::port::subscriber::Subscriber`]s. If the [`SampleMut`] is not sent
+/// it will release the loaned memory when going out of scope.
+pub trait SampleMut<MessageType>: internal::SampleMgmt {
```

Comment:
Since this is coupled to the `PublisherLoan` trait it makes more sense to have a `PublisherLoanUninit` trait for the `UninitializedSampleMut` struct